test/test_pipeline.py

```diff
@@ -0,0 +1,247 @@
+    assert "hb" in adata.var.columns
+    assert "n_counts" in adata.var.columns
+    assert "n_cells" in adata.var.columns
+
+
+def test_generate_qc_clusters():
+    # create test data
+    adata = sc.datasets.pbmc68k_reduced()
+
+    # test that generate_qc_clusters adds expected columns to adata.obs and adata.obsm
+    calculate_qc(adata)
+    generate_qc_clusters(
+        adata,
+        [
+            "n_counts",
+            "n_genes",
+            "percent_mito",
+            "percent_ribo",
+            "percent_hb",
+        ],
+    )
+    assert "qc_cluster" in adata.obs.columns
+    assert "X_umap_qc" in adata.obsm.keys()
+
+
+def test_scale_factor():
+    # test that _scale_factor returns expected values
+    assert _scale_factor(np.array([0, 1, 2, 3, 4, 5])) == 1.0
+    assert _scale_factor(np.array([0, 1, 2, 3, 10])) == 0.5
+
+
+def test_fit_gaussian():
+    # create test data (normal distribution centered on 5)
+    x = np.random.normal(5, 2, 1000)
+
+    # test that fit_gaussian returns expected values
+    x_peak, x_left, x_right = fit_gaussian(x, n=2, threshold=0.05, plot=False)
+    assert -1 < x_peak < 1
+    assert 9 < x_left < 11
+    assert isinstance(x_right, GaussianMixture)
+
+
+def test_filter_qc_outlier2():
+    # Load example dataset
+    adata = sc.datasets.pbmc68k_reduced()
+
+    # Test default metrics
+    pass_filter = filter_qc_outlier2(adata)
+    assert isinstance(pass_filter, np.ndarray)
+    assert pass_filter.shape[0] == adata.shape[0]
+
+    # Test custom metrics
+    pass_filter = filter_qc_outlier2(
+        adata,
+        metrics={
+            "n_counts": [500, None, "log", "min_only", 0.5],
+            "percent_mito": [0.05, None, "linear", "max_only", 0.5],
+        },
+    )
+    assert isinstance(pass_filter, np.ndarray)
+    assert pass_filter.shape[0] == adata.shape[0]
+
+    # Test force=True
+    pass_filter = filter_qc_outlier2(
+        adata,
+        metrics=["n_counts", "percent_mito"],
+        force=True,
+    )
+    assert isinstance(pass_filter, np.ndarray)
+    assert pass_filter.shape[0] == adata.shape[0]
+
+
+def test_find_good_qc_cluster():
+    # Load example dataset
+    adata = sc.datasets.pbmc3k()
+    calculate_qc(adata)
+    generate_qc_clusters(
+        adata,
+        [
+            "n_counts",
+            "n_genes",
+            "percent_mito",
+            "percent_ribo",
+            "percent_hb",
+        ],
+    )
+
+    # Test default metrics
+    find_good_qc_cluster(adata)
+    assert "fqo2" in adata.obs
+    assert "good_qc_clusters" in adata.obs
+
+    # Test custom metrics (all samples fail here)
+    metrics = {
+        "n_counts": (500, 5000, "log", "both", 0.1),
+        "n_genes": (200, 5000, "log", "both", 0.1),
+        "percent_mito": (0, 1, "log", "both", 0.1),
+    }
+    find_good_qc_cluster(
+        adata,
+        metrics=metrics,
+        threshold=0.6,
+        key_added="custom_good_qc_clusters",
+    )
+    assert "fqo2" in adata.obs
```

Comment:
the `key_added` is used when constructing the output `.obs`, and this is actually under `"custom_good_qc_clusters_fqo2"`

Reply:
I rejiggered the code a bit on my end and this will need a different change now

Reply:
do you want to commit your changes to the PR, or is this just a note to yourself?

Reply:
I'm putting finishing touches on a demo notebook and will push some stuff relatively soon. Sorry about this, should have been clearer, but messing around with things in practice gave me a couple ideas and I let things flow.